feat(frontend): add feature introduction dialog with onboarding flow and update footer info

ericsen-tsai opened the pull request:
## Summary

This PR introduces a new feature introduction dialog system that provides an onboarding experience for users. The dialog displays a multi-step guide (4 steps) explaining the new "呼叫報導仔" (Call Baodaozai) feature and related functionality. The dialog automatically appears for first-time users and can be manually triggered through baodaozai event triggers.

## Changes

- **New Feature Intro Dialog Component**: Created a multi-step dialog component (`feature-info-dialog`) with step navigation, indicator dots, and responsive image support
- **Feature Intro Service**: Implemented a new service module (`services/feature-intro/`) with:
  - Context provider (`FeatureIntroDialogProvider`) for managing dialog state across the app
  - Configuration file with 4 onboarding steps in Traditional Chinese
  - Utility functions for rendering markdown-style bold text in descriptions
  - Constants for localStorage keys to track if user has seen the dialog
- **Onboarding Images**: Added 12 SVG images (3 sizes × 4 steps) for responsive display:
  - Small (`s`): Mobile viewport
  - Medium (`m`): Tablet viewport  
  - Large (`l`): Desktop viewport
- **Integration with Baodaozai Event Triggers**: 
  - Updated `all-site-baodaozai-event-trigger.tsx` to integrate with feature intro dialog context
  - Updated `start-reading-baodaozai-event-trigger.tsx` to respect dialog visibility state
  - Baodaozai triggers now only show when the feature intro dialog has been seen or closed
- **Layout Integration**: Added `FeatureIntroDialog` component to root layout for global availability
- **Provider Setup**: Added `FeatureIntroDialogProvider` to the providers tree
- **Package Updates**: 
  - Updated `@kids-reporter/routing-ui` from `0.1.1` to `0.1.2`
  - Updated `routing-ui` package version in its own `package.json`
- **Footer Information Update**: Update the information of footer

## Additional Notes

- The dialog uses localStorage to persist whether a user has seen the introduction (`feature-intro-dialog-seen` key)
- Dialog automatically opens on first page load if the user hasn't seen it before
- Users can navigate through 4 steps or close the dialog at any time
- The dialog is responsive and adapts its layout and images based on screen size
- Integration ensures baodaozai features are only shown after users have been introduced to the feature
- This PR is based on #863 

## Screenshot(s)

https://github.com/user-attachments/assets/98082f6b-4f23-478e-aafc-fe045d39e360

## Reference(s)
- [[前台] 全站功能介紹 - UI](https://www.notion.so/twreporter/UI-2cd8dbdca93280fbba59e26bb6d9ca2f?source=copy_link)
- [[前台] ndx footer 更新](https://www.notion.so/twreporter/ndx-footer-2cd8dbdca932809593ecdc2992f6f150?source=copy_link)

ericsen-tsai committed:
feat(frontend): add onboarding SVG images for user guidance

ericsen-tsai committed:
feat(frontend): implement feature introduction dialog and context

- Added FeatureIntroDialog component to guide users through new features with interactive steps.
- Introduced FeatureIntroDialogProvider for managing dialog state and visibility.
- Created configuration for feature introduction content and utility for rendering descriptions.
- Added indicator dots for navigation between steps in the dialog.
- Integrated the dialog into the main layout and updated Providers to include the new context.

ericsen-tsai committed:
feat(frontend): integrate feature intro dialog context into event triggers

- Added `canShowBaodaozai` logic to determine visibility of the Baodaozai event triggers based on the feature intro dialog state.
- Updated `StartReadingBaodaozaiEventTrigger` and `AllSiteBaodaozaiEventTrigger` components to utilize the new context for enabling/disabling triggers.
- Refactored state management in the `FeatureIntroDialogProvider` to support the new feature visibility logic.

ericsen-tsai committed:
feat(routing-ui): update footer content info

ericsen-tsai committed:
chore: update routing-ui package version

ericsen-tsai commented on packages/frontend/src/services/feature-intro/utils.tsx:
```diff
@@ -0,0 +1,14 @@
+export const renderDescription = (text: string) => {
+  const parts = text.split(/(\*\*.*?\*\*)/g)
+  return parts.map((part, index) => {
+    if (part.startsWith('**') && part.endsWith('**')) {
+      const boldText = part.slice(2, -2)
+      return (
+        <strong key={index} className="font-bold">
+          {boldText}
+        </strong>
+      )
+    }
+    return <span key={index}>{part}</span>
+  })
+}
```
Since the description is from the config, the edge case should be aware of.
I'll handle it if the wording matches the edge case.

ericsen-tsai committed:
fix(frontend): update feature intro dialog context and accessibility

- Refactored `canShowBaodaozai` logic in `FeatureIntroDialogProvider` to check for `window` before accessing `localStorage`.
- Added `aria-label` to the button in `FeatureIntroDialog` for improved accessibility, indicating the action based on the current step.

ericsen-tsai commented on packages/frontend/src/services/feature-intro/components/feature-info-dialog/index.tsx:
```diff
@@ -0,0 +1,107 @@
+          </div>
+
+          <div className="mt-auto flex items-center justify-center gap-8 desktop:justify-start">
+            <IndicatorDots current={step} total={totalSteps} />
+            <Button
+              variant="primary"
+              size={44}
+              onClick={handleNext}
+              className="w-30 desktop:order-[-1]"
+            >
+              {step === totalSteps ? '完成' : '下一步'}
+            </Button>
```
Fixed in [fix(frontend): update feature intro dialog context and accessibility](https://github.com/kids-reporter/kids-reporter-monorepo/pull/864/changes/704eedb0e2bdf974e9bd82a3605d2b4e24269d6a)

ericsen-tsai commented on packages/frontend/src/services/feature-intro/context.tsx:
```diff
@@ -0,0 +1,60 @@
+      canShowBaodaozai:
+        (isDialogOpen !== undefined && !isDialogOpen) ||
+        (typeof localStorage !== 'undefined' &&
+          localStorage.getItem(FEATURE_INTRO_DIALOG_SEEN_KEY) === 'true'),
```
Fixed in [fix(frontend): update feature intro dialog context and accessibility](https://github.com/kids-reporter/kids-reporter-monorepo/pull/864/changes/704eedb0e2bdf974e9bd82a3605d2b4e24269d6a)

nickhsine commented on packages/frontend/src/app/(sticky-header)/_components/article/start-reading-baodaozai-event-trigger.tsx:
```diff
@@ -13,6 +15,14 @@ function StartReadingBaodaozaiEventTrigger({
+      return true
+    }
+    return !isFirstRenderAtTop
+  }, [canShowBaodaozai, isFirstRenderAtTop])
```
nitpick:
好像不需要用 useMemo?

ericsen-tsai replied:
Yes, adjusted in [fix(frontend): update feature intro dialog context and event trigger logic](https://github.com/kids-reporter/kids-reporter-monorepo/pull/864/changes/91a22d6f648f21e3d7ef9f5dfe8bec5f19f4752a)

nickhsine commented on packages/frontend/src/components/all-site-baodaozai-event-trigger.tsx:
```diff
@@ -75,6 +75,15 @@ function AllSiteBaodaozaiEventTrigger({
+      return true
+    }
+    return id === 'show-intro' && !isFirstRenderAtTop
+  }, [canShowBaodaozai, id, isFirstRenderAtTop])
```
nitpick:
同上，好像不用 useMemo

ericsen-tsai replied:
Adjusted in [fix(frontend): update feature intro dialog context and event trigger logic](https://github.com/kids-reporter/kids-reporter-monorepo/pull/864/changes/91a22d6f648f21e3d7ef9f5dfe8bec5f19f4752a)

nickhsine commented on packages/frontend/src/services/feature-intro/components/feature-info-dialog/index.tsx:
```diff
@@ -0,0 +1,107 @@
+  const currentImageSrc = useMemo(() => {
+    const size = isTablet ? 'm' : isDesktop ? 'l' : 's'
+    return `/assets/images/feature-intro/onboarding_${step}_${size}.svg`
+  }, [isDesktop, isTablet, step])
```
nitpick:
似乎不需要 useMemo。

ericsen-tsai replied:
Adjusted in [fix(frontend): update feature intro dialog context and event trigger logic](https://github.com/kids-reporter/kids-reporter-monorepo/pull/864/changes/91a22d6f648f21e3d7ef9f5dfe8bec5f19f4752a)

nickhsine commented on packages/frontend/src/services/feature-intro/components/feature-info-dialog/index.tsx:
```diff
@@ -0,0 +1,107 @@
+  const isDesktop = useMediaQuery('(min-width: 1024px)')
+
+  const currentImageSrc = useMemo(() => {
+    const size = isTablet ? 'm' : isDesktop ? 'l' : 's'
```
issue:
目前的判斷會讓 1025px （>= 1024px）螢幕寬度，使用 'm' 的 size。
照理來說，應該是要用 'l' size。

ericsen-tsai replied:
Fixed in [fix(frontend): update feature intro dialog context and event trigger logic](https://github.com/kids-reporter/kids-reporter-monorepo/pull/864/changes/91a22d6f648f21e3d7ef9f5dfe8bec5f19f4752a)

nickhsine commented on packages/frontend/src/services/feature-intro/context.tsx:
```diff
@@ -0,0 +1,51 @@
+      openDialog,
+      closeDialog,
+    }),
+    [isDialogOpen, openDialog, closeDialog]
```
nitpick:
好像不需要特別使用 useCallback 和 useMemo ？
因為就這個情況來說，好像沒有大量的計算，所以也不需要 stable reference。

ericsen-tsai replied:
I think I'll stick with `useMemo`. Without it, a new object is created on every render, causing all context consumers to re-render even when values haven't changed.

I made a refactor to rename the `canShowBaodaozai` to `isFinishedIntro` and also memorized this value separately for complicated computation.

[fix(frontend): refactor feature intro dialog context and component logic](https://github.com/kids-reporter/kids-reporter-monorepo/pull/864/changes/52a73745755e5891c1869198fb2702eff1d88a1b)

nickhsine reviewed (approved):
LGTM 💯 
有幾個 nitpick 和一個 issue 再麻煩看看。

ericsen-tsai committed:
fix(frontend): update feature intro dialog context and event trigger logic

- Replaced `canShowBaodaozai` with `isFinishedIntro` in `FeatureIntroDialogContext` to streamline the logic for enabling/disabling event triggers.
- Simplified the `disabled` state calculations in `StartReadingBaodaozaiEventTrigger` and `AllSiteBaodaozaiEventTrigger` components.
- Adjusted the image source logic in `FeatureIntroDialog` for improved readability.

ericsen-tsai committed:
fix(frontend): refactor feature intro dialog context and component logic

- Memoized `isFinishedIntro` value in `FeatureIntroDialogProvider`.
- Updated the `FeatureIntroDialog` component to ensure the `hasSeenDialog` check directly compares the localStorage value to 'true' for clarity.
